File: cpp/567. Permutation in String.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

/*
How to check whether one string is a permutation of another string?
1. Sort the string and then compare
2. Both strings must have same character frequencies
*/

class Solution {
// Sliding Window + HashMap
public:
    bool checkInclusion(string s1, string s2) {
        int windowSize = s1.size(), n = s2.size();
        if(windowSize > n) return false;

        unordered_map<char, int> s1Counter, s2Counter;
        for(int i = 0; i < windowSize; i++) {
            s1Counter[s1[i]] += 1;
            s2Counter[s2[i]] += 1;
        }
        if(s1Counter==s2Counter) return true;

        for(int l = 1; l <= n-windowSize; l++) {
            int r = l + windowSize - 1;
            s2Counter[s2[l-1]] -= 1;
            if(s2Counter[s2[l-1]] == 0) s2Counter.erase(s2[l-1]); 
            s2Counter[s2[r]] += 1;
            if(s1Counter==s2Counter) return true;
        }
        return false;
    }
};

class Solution {
// Sliding Window + Array
public:
    bool checkInclusion(string s1, string s2) {
        int windowSize = s1.size(), n = s2.size();
        if(windowSize > n) return false;

        int s1Counter[26] = {0}, s2Counter[26] = {0};
        for(int i = 0; i < windowSize; i++) {
            s1Counter[s1[i]-'a'] += 1;
            s2Counter[s2[i]-'a'] += 1;
        }
        if(equal(s1Counter, s1Counter+26, s2Counter)) return true;

        for(int l = 1; l <= n-windowSize; l++) {
            int r = l + windowSize - 1;
            s2Counter[s2[l-1]-'a'] -= 1;
            s2Counter[s2[r]-'a'] += 1;
            if(equal(s1Counter, s1Counter+26, s2Counter)) return true;
        }
        return false;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; string s1, s2; bool ans;

    s1 = "ab", s2 = "eidbaooo";
    ans = solution.checkInclusion(s1, s2);
    assert(ans==true);

    s1 = "ab", s2 = "eidboaoo";
    ans = solution.checkInclusion(s1, s2);
    assert(ans==false);
}

File: python/567-Permutation in String.py
from collections import Counter

'''
How to check whether one string is a permutation of another string?
1. Sort the string and then compare
2. Both strings must have same character frequencies
https://leetcode.com/problems/permutation-in-string/solution/
'''

class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        "Approach2: Sort, Time: O(l1logl1 + (l2-l1)*(l1logl1+l1)), Space: O(1), TLE"
        s1 = sorted(s1) # O(l1logl1)
        window_sz = len(s1)
        for i in range(len(s2)-window_sz+1): # O(l2-l1)
            sub_s2 = sorted(s2[i:i+window_sz]) # O(l1logl1)
            if s1 == sub_s2: return True # O(l1)
        return False
    def checkInclusion(self, s1: str, s2: str) -> bool:
        "Approach3: HashTable, Time: O(l1+(l2-l1)*l1), Space: O(1)"
        counter1 = Counter(s1) # O(l1)
        window_sz = len(s1)
        for i in range(len(s2)-window_sz+1): # O(l2-l1)
            counter2 = Counter(s2[i:i+window_sz])  # O(l1)
            if counter1 == counter2: return True # O(26)
        return False
    def checkInclusion(self, s1: str, s2: str) -> bool:
        "Approach5 but use Counter: HashTable + Sliding Window, Time: O(l1+(l2-l1)), Space: O(1)"
        counter1 = Counter(s1) # O(l1)
        window_sz = len(s1)

        counter2 = Counter(s2[:window_sz])
        if counter1 == counter2: return True
        for i in range(1, len(s2)-window_sz+1): # O(l2-l1)
            counter2[s2[i-1]] -= 1
            if counter2[s2[i-1]] <= 0: del counter2[s2[i-1]]
            counter2[s2[i+window_sz-1]] += 1
            if counter1 == counter2: return True # O(26)
        return False
    def checkInclusion(self, s1: str, s2: str) -> bool:
        "Approach5: Array + Sliding Window, Time: O(l1+(l2-l1)), Space: O(1)"
        if len(s1) > len(s2): return False
        window_sz = len(s1)
        counter1, counter2 = [0]*26, [0]*26
        for i in range(window_sz): # O(l1)
            counter1[ord(s1[i]) - ord('a')] += 1
            counter2[ord(s2[i]) - ord('a')] += 1
        if counter1 == counter2: return True
        for i in range(1, len(s2)-window_sz+1): # O(l2-l1)
            counter2[ord(s2[i-1]) - ord('a')] -= 1
            counter2[ord(s2[i+window_sz-1]) - ord('a')] += 1
            if counter1 == counter2: return True # O(26)
        return False

s1 = "ab"
s2 = "eidbaooo"
result = Solution().checkInclusion(s1, s2)
assert result == True

s1= "ab"
s2 = "eidboaoo"
result = Solution().checkInclusion(s1, s2)
assert result == False